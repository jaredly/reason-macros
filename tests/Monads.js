// Generated by BUCKLESCRIPT VERSION 4.0.1100, PLEASE EDIT WITH CARE
'use strict';

var Pervasives = require("bsb-native/lib/js/pervasives.js");

var match = 3;

var x = match !== undefined ? Pervasives.failwith("Unwrapped optional") : undefined;

exports.x = x;
/* x Not a pure module */
