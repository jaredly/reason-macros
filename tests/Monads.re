
let%macro.let opt = (pattern, value, continuation) => {
  switch eval__value {
    | Some(eval__pattern) => eval__continuation
    | None => None
  }
};

// can also do
[@macro.name "opt.force"]
let%macro.let _ = (pattern, value, continuation) => {
  switch eval__value {
    | Some(eval__pattern) => eval__continuation
    | None => failwith("Unwrapped optional")
  }
};

let x = {
  let%opt a = Some(3);
  let d = 4;
  let%opt.force () = None;
  Some(a + d)
};