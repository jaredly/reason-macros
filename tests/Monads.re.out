let x = {
  switch (Some(3)) {
  | Some(a) =>
    let d = 4;
    switch (None) {
    | Some () => Some(a + d)
    | None => failwith("Unwrapped optional")
    };
  | None => None
  };
};
